src/core.jl

```diff
@@ -282,6 +282,45 @@ Base.similar(A::AxisArray, ::Type{S}, ax1::Axis, axs::Axis...) where {S} = simil
     end
 end
 
+Base.convert(::Type{AxisArray{T, N, A, As}},
+             array::AxisArray{T, N, A, As}) where {T, N, A, As} = array
+Base.convert(::Type{AxisArray{TT, N, AA, As}},
+             array::AxisArray{T, N, A, As}) where {TT, N, AA, As, T, A} =
+    AxisArray(convert(AA, array), axes(array))
+function Base.convert(I::Type{AxisArray{TT, N, AA, AAs}},
+                      array::AxisArray{T, N, A, As}
+                     )::AxisArray{TT, N, AA, AAs} where {TT, N, AA, AAs, T, A, As}
+    perm = map(x -> findfirst(y -> typeof(y) <: x, axes(array)),
```

Comment:
This line will cause conversion to fail when the second type parameter of `Axis` differs, e.g. `Base.OneTo` vs. `Base.UnitRange`:

```
julia> A = AxisArray(reshape(1:24, 2, 3, 4), Axis{:a}(1:2), Axis{:b}(1:3), Axis{:c}(1:4))
3-dimensional AxisArray{Int64,3,...} with axes:
    :a, 1:2
    :b, 1:3
    :c, 1:4
And data, a 2×3×4 Base.ReshapedArray{Int64,3,UnitRange{Int64},Tuple{}}:
[:, :, 1] =
 1  3  5
 2  4  6

[:, :, 2] =
 7   9  11
 8  10  12

[:, :, 3] =
 13  15  17
 14  16  18

[:, :, 4] =
 19  21  23
 20  22  24

julia> Adense = AxisArray(zeros(Int16, size(A)), :c, :a, :b)
3-dimensional AxisArray{Int16,3,...} with axes:
    :c, Base.OneTo(2)
    :a, Base.OneTo(3)
    :b, Base.OneTo(4)
And data, a 2×3×4 Array{Int16,3}:
[:, :, 1] =
 0  0  0
 0  0  0

[:, :, 2] =
 0  0  0
 0  0  0

[:, :, 3] =
 0  0  0
 0  0  0

[:, :, 4] =
 0  0  0
 0  0  0

julia> convert(typeof(Adense), A)
ERROR: ArgumentError: Axes do not match
Stacktrace:
 [1] convert(::Type{AxisArrays.AxisArray{Int16,3,Array{Int16,3},Tuple{AxisArrays.Axis{:c,Base.OneTo{Int64}},AxisArrays.Axis{:a,Base.OneTo{Int64}},AxisArrays.Axis{:b,Base.OneTo{Int64}}}}}, ::AxisArrays.AxisArray{Int64,3,Base.ReshapedArray{Int64,3,UnitRange{Int64},Tuple{}},Tuple{AxisArrays.Axis{:a,UnitRange{Int64}},AxisArrays.Axis{:b,UnitRange{Int64}},AxisArrays.Axis{:c,UnitRange{Int64}}}}) at /Users/ajkeller/.julia/v0.6/AxisArrays/src/core.jl:295
```

If I understand correctly what this method is intended for, it seems like it would be better to compare axis names instead of axis types. The functions `axisnames` and `AxisArrays.axisname` (unexported) may be useful for this.

Reply:
Just getting back into this after some time away. Conversions between arrays using just the name is a more interesting problem. More specifically, I now have to convert axes as well since their types may differ. I'll have to think about this a bit more, but I may need to include axis conversion functions to do this properly.